### ISSUE TITLE
hdnn.py distances not correct for periodically repeated bulk structure

### ISSUE THREAD

**Issue by lmj1029123** [atomistic-machine-learning/schnetpack#104]
Hi,

It seems that the distances calculated in the hdnn.py is not correct. 

```
 # Compute radial functions
        if self.RDF is not None:
            # Get atom type embeddings
            Z_rad = self.radial_Z(Z)
            # Get atom types of neighbors
            Z_ij = snn.neighbor_elements(Z_rad, neighbors)
            # Compute distances
            distances = snn.atom_distances(
                positions, neighbors, neighbor_mask=neighbor_mask
            )
            radial_sf = self.RDF(
                distances, elemental_weights=Z_ij, neighbor_mask=neighbor_mask
            )
        else:
            radial_sf = None
```
If I have a Cu bulk with a single Cu atom repeated in all 3 directions, the distances vector looks like [0, 0, 0,....,0]. I believe it should be [a, a, a, ..... , 2a etc] if a is the lattice constant.  This is probably because the cell_offset is not used in the calculation? Could you have a look at it? Thanks!

**Comment by lmj1029123** [comment]
Hi,

Just to let you know that I am still waiting for the solution. Thank you very much.

**Comment by mgastegger** [comment]
Hi,
sorry for the long wait. I'm currently looking into it. The distances should be easy to fix, the angles proof to be a little trickier.

**Comment by lmj1029123** [comment]
Another thing I noticed is that when I use the Behler representation, with radial function only, I get nan for loss for all the time. It doesn't matter how I set the learning rate. 
```
import schnetpack.output_modules
from schnetpack.data import AtomsData
import pickle
import json
import torch
import torch.nn.functional as F
from torch.optim import Adam

import schnetpack as spk
import schnetpack.atomistic as atm
import schnetpack.representation as rep
from schnetpack.datasets import *
from schnetpack.metrics import MeanSquaredError
from schnetpack.environment import AseEnvironmentProvider

config = {'Database':{'Name': 'BPNN_original_300_diag',
                      'datafile': './db/master-jb.db',
                      'train_size': 300,
                      'val_size': 100,
                      'batch_size': 5},
          'Fingerprints':{'n_radial': 4,
                          'n_angular': 0,
                          'cutoff_radius': 6,
                          'mode': 'Behler'},
          'Model':{'Type': 'BPNN',
                   'n_layers': 2,
                   'n_hidden': 100,
                   'lr': 1e-4,
                   'E_coef': 1,
                   'F_coef': 1,
                   'val_interval': 1,
                   'n_epochs': 1000}}


Name = config['Database']['Name']
datafile = config['Database']['datafile']
train_size = config['Database']['train_size']
val_size = config['Database']['val_size']
batch_size = config['Database']['batch_size']
n_radial = config['Fingerprints']['n_radial']
n_angular = config['Fingerprints']['n_angular']
cutoff_radius = config['Fingerprints']['cutoff_radius']
mode = config['Fingerprints']['mode']
n_layers = config['Model']['n_layers']
n_hidden = config['Model']['n_hidden']
lr = config['Model']['lr']
E_coef = config['Model']['E_coef']
F_coef = config['Model']['F_coef']
val_interval = config['Model']['val_interval']
n_epochs = config['Model']['n_epochs']



if not os.path.exists(Name):
    os.makedirs(Name)

config_name = f'./{Name}/config.json'
with open(config_name,'w') as f:
    json = json.dump(config,f)

#config_file = open(config_name,'w')
#config_file.write(json)
#config_file.close()

data = AtomsData(datafile, required_properties=['energy','forces','stress'],collect_triples=True,environment_provider = AseEnvironmentProvider(cutoff_radius))


train, val, test = data.create_splits(train_size, val_size)
train_loader = spk.data.AtomsLoader(train, batch_size=batch_size, num_workers=0)
val_loader = spk.data.AtomsLoader(val)
test_loader = spk.data.AtomsLoader(test)
loader = [train_loader, val_loader, test_loader]
pickle.dump(loader, open('./'+Name+'/loader.sav','wb'))


reps = rep.BehlerSFBlock(n_radial=n_radial, n_angular=n_angular, elements=frozenset([46,79]), cutoff_radius=cutoff_radius, mode = mode)
reps = rep.StandardizeSF(reps, train_loader)
mean, stddev = train_loader.get_statistics('energy', True)


# Should be noticed that I add a "negative_dr" argument here. I did this to make sure the force is the negative gradient. I also modified the code in output_module to accommodate that. 
output = schnetpack.output_modules.ElementalAtomwise(n_in=reps.n_symfuncs,n_layers=n_layers,n_hidden=n_hidden, elements=frozenset([46,79]),property='energy',derivative='forces',negative_dr=True,create_graph=True,mean=mean['energy'],stddev=stddev['energy'])
model = atm.AtomisticModel(reps, output,elements=frozenset([46,79]))

trainable_params = filter(lambda p: p.requires_grad, model.parameters())

metric_E = [MeanSquaredError(target = 'energy',model_output='energy'),MeanSquaredError(target = 'forces',model_output='forces')]
hook_E = spk.train.CSVHook("./"+Name, metric_E)

opt = Adam(trainable_params, lr=lr)

loss = lambda b, p: E_coef*F.mse_loss(p["energy"], b['energy'])+F_coef*F.mse_loss(p["forces"], b['forces'])
trainer = spk.train.Trainer(Name+"/", model, loss,
                            opt, train_loader, val_loader,hooks = [hook_E], validation_interval = val_interval)

# start training
trainer.train(torch.device("cpu"),n_epochs = n_epochs)
```
If you don't notice that problem, it will be great if you could show me the set up for running the code. Thank you very much!

**Comment by bencomer** [comment]
Thanks for working on this problem @mgastegger -- our group is also very interested in seeing this fixed.

**Comment by mgastegger** [comment]
Just an update, distances are running properly, at least for energies. I will check the forces next, to see whether they are the source of the NaNs. I'm also working on the angular symmetry functions.

**Comment by lmj1029123** [comment]
Hi, 

I believe the NaN problem is not in the forces. They were probably from the fingerprints. I wrote radial fingerprint function. While it is kind of slow, it does not give me the NaN problem. Also, I am wondering where I can get the the updated distances. It will be great to have a model where the radial symmetry could work, even if the angular symmetry function is not working yet. Thanks!

**Comment by mgastegger** [comment]
Thanks for the information.
Unfortunately, I was sidetracked by a few submission deadlines until now.
I will push a version with the distances fixed on Monday.

**Comment by lmj1029123** [comment]
Sounds good. Thanks for working on it!

**Comment by mgastegger** [comment]
Radial SFs should work with PBC now. The angular ones will take more time.

@lmj1029123: Unfortunately, I was not able to reproduce the NaN issue using the materials project data and your script. Does the problem arise immediately, or only after a time? You could also disabling the StandardizeSF layer to see, whether this is the issue.

**Comment by lmj1029123** [comment]
> Radial SFs should work with PBC now. The angular ones will take more time.
> 
> @lmj1029123: Unfortunately, I was not able to reproduce the NaN issue using the materials project data and your script. Does the problem arise immediately, or only after a time? You could also disabling the StandardizeSF layer to see, whether this is the issue.

The problem arise immediately. However, the NaN issue actually goes away with PBC fix. Thanks!

**Comment by mgastegger** [comment]
Great to hear, I'll keep you updated on the angular functions.

**Comment by lmj1029123** [comment]
Hi,

I think I have a fix for the angular symmetry functions and I made a pull request for it. It seems that it is really expensive as it looks at N_atoms^2 neighbor pairs. I am not sure whether my implementation is compatible with GPU, Also, I am wondering whether there is tutorial of how to use GPU for schnetpack? Thanks!

**Comment by mgastegger** [comment]
Thank you very much, that was a huge help!
I went through the code and ran a few tests. It works on GPU and CPU and appears to be stable for gradients. The scaling of the angular functions is always a problem. Avoiding this is one of the advantages of SchNet. 

For running on GPU it should be enough to change your example script to contain:
`device = torch.device('cuda')`
`model = atm.AtomisticModel(reps, output,elements=frozenset([46,79])).to(device)`

Sorry again for the delay, we were quire busy but now things should be back to normal.

**Comment by ktschuett** [comment]
PR #110 is merged.
